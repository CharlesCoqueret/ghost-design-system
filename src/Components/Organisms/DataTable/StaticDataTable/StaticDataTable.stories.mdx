import { Canvas, Meta, ArgsTable, Story, Source } from '@storybook/addon-docs';

import StaticDataTable from './StaticDataTable';
import { ColumnType } from '../Common/types';
import FakeExtraStaticDataTable from '../Common/FakeComponents/FakeExtraStaticDataTable';
import { Icon } from '../../../Atoms';

import dataTableStyle from '../../../../assets/_datatable.scss?raw';

export const TemplateStaticDataTable = (args) => {
  return <StaticDataTable {...args} />;
};

<Meta
  title='Organism/DataTable'
  component={StaticDataTable}
  parameters={{
    viewMode: 'docs',
    layout: 'fullscreen',
  }}
/>

# StaticDataTable

The data table displays information in a grid-like format of rows and columns. They organize information in an easy to read and scan through way.

Play with StaticDataTable live on [CodeSandbox](https://codesandbox.io/s/datatable-demo-z5fb38)

## Props

<ArgsTable of={StaticDataTable} sort='requiredFirst' exclude={['dataTestId']} />

## extra: IExtraStaticDataTableProps<T\>

<ArgsTable of={FakeExtraStaticDataTable} sort='requiredFirst' />

## Story

<Canvas>
  <Story
    name='StaticDataTable'
    args={{
      data: [
        {
          id: 'UGA',
          name: 'Lorem ipsum dolor sit amet consectetur adipiscing elit',
          status: 'INACTIVE',
          price: -100000,
          parts: -10,
          startDate: new Date(2021, 2, 24),
        },
        {
          id: 'ARE',
          name: 'Consecte',
          status: 'ACTIVE',
          price: -2500000,
          parts: 15,
          startDate: new Date(2021, 1, 13),
        },
        {
          id: 'GBR',
          name: 'Sed',
          status: 'IN_PROGRESS',
          price: 150000,
          parts: 3,
          startDate: new Date(2021, 1, 4),
        },
        {
          id: 'JUF',
          name: 'dolore eu fugiat nulla pariatur',
          status: 'IN_PROGRESS',
          price: -354000,
          parts: 70,
          startDate: new Date(2021, 1, 15),
        },
        {
          id: 'YAI',
          name: 'Sed ut perspiciatis',
          status: 'ACTIVE',
          price: 4150000,
          parts: 30,
          startDate: new Date(2021, 2, 17),
        },
        {
          id: 'SML',
          name: 'Ut enim ad minima veniam',
          status: 'ACTIVE',
          price: -70500,
          parts: 10,
          startDate: new Date(2021, 3, 10),
        },
      ],
      columns: [
        {
          title: 'Code',
          dataIndex: 'id',
          sorter: true,
          type: ColumnType.CODE,
          width: '70px',
        },
        {
          title: 'Text',
          dataIndex: 'name',
          type: ColumnType.TEXT,
          ellipsis: true,
          sorter: true,
          width: '20%',
        },
        {
          title: 'Badge',
          dataIndex: 'status',
          sorter: true,
          type: ColumnType.BADGE,
          options: [
            { label: 'Inactive', value: 'INACTIVE' },
            { label: 'In progress', value: 'IN_PROGRESS' },
            { label: 'Active', value: 'ACTIVE' },
          ],
        },
        {
          title: 'Amount',
          dataIndex: 'price',
          sorter: true,
          type: ColumnType.AMOUNT,
          currency: 'â‚¬',
        },
        {
          title: 'Percentage',
          dataIndex: 'parts',
          sorter: true,
          type: ColumnType.PERCENTAGE,
        },
        {
          title: 'Date',
          dataIndex: 'startDate',
          sorter: true,
          type: ColumnType.DATE,
        },
        {
          title: 'Actions',
          type: ColumnType.BUTTON,
          moreActionsMessage: 'Nore actions',
          buttons: [
            {
              icon: ['fal', 'arrow-down-to-line'],
              label: 'Download',
              onClick: (row) => {
                console.log(`download icon clicked on row: ${row.id}`);
              },
            },
            {
              icon: ['fal', 'eye'],
              label: 'Watch',
              onClick: (row) => {
                console.log(`Eye icon clicked on row: ${row.id}`);
              },
            },
          ],
        },
      ],
      extra: {
        computeTotal: (data, dataIndex) => {
          if (dataIndex === 'price' || dataIndex === 'parts') {
            return data
              .map((row) => {
                return row[dataIndex];
              })
              .reduce((a, b) => Number(a) + Number(b), 0);
          } else {
            return '-';
          }
        },
        onRowClick: (row) => {
          console.log(`clicked row: ${row.id}`);
        },
        isSelectable: (row) => {
          return row.id !== 'UGA';
        },
        onRowSelect: (rows, row) => {
          console.log(`Number of rows selected: ${rows.length}\nSelected row: ${row.id}`);
        },
      },
    }}>
    {TemplateStaticDataTable.bind({})}
  </Story>
</Canvas>
