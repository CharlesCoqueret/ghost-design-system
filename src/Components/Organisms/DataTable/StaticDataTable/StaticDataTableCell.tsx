import React, { ReactElement } from 'react';

import { ColumnType, IColumnType, IExtraStaticDataTableProps } from '../Common/types';
import {
  AmountCell,
  BadgeCell,
  ButtonCell,
  CheckboxCell,
  CodeCell,
  CustomCell,
  DynamicSearchCell,
  DateCell,
  NumberCell,
  PercentageCell,
  TextCell,
  TextAreaCell,
} from '../Common/Cells';

export interface IStaticDataTableCellProps<T> {
  column: IColumnType<T>;
  dataTestId?: string;
  row: T;
  extra?: IExtraStaticDataTableProps<T>;
  rowIndex: number;
}

const StaticDataTableCell = <T,>(props: IStaticDataTableCellProps<T>): ReactElement => {
  const { column, dataTestId, row, extra, rowIndex } = props;

  switch (column.type) {
    case ColumnType.AMOUNT: {
      return (
        <AmountCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.BADGE: {
      return (
        <BadgeCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.BUTTON: {
      return (
        <ButtonCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.CHECKBOX: {
      return (
        <CheckboxCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.CODE: {
      return (
        <CodeCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.CUSTOM: {
      return (
        <CustomCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.DATE: {
      return (
        <DateCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.DYNAMICSEARCH: {
      return (
        <DynamicSearchCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.NUMBER: {
      return (
        <NumberCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.PERCENTAGE: {
      return (
        <PercentageCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.TEXT: {
      return (
        <TextCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.TEXTAREA: {
      return (
        <TextAreaCell<T>
          column={column}
          dataTestId={dataTestId ? `${dataTestId}-${column.title}-${rowIndex}` : undefined}
          row={row}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    default: {
      throw new Error('Missing ColumnType');
    }
  }

  throw new Error('Should have returned by then');
};

export default StaticDataTableCell;
