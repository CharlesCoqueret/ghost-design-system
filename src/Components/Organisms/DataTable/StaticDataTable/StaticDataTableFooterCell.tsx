import React, { ReactElement } from 'react';

import { ColumnType, IColumnType, IExtraStaticDataTableProps } from '../Common/types';
import { AmountCell, NumberCell, PercentageCell } from '../Common/Cells';

export interface IStaticDataTableFooterCellProps<T> {
  column: IColumnType<T>;
  data: Array<T>;
  extra?: IExtraStaticDataTableProps<T>;
  rowIndex: number;
}

const StaticDataTableFooterCell = <T,>(props: IStaticDataTableFooterCellProps<T>): ReactElement => {
  const { column, data, extra, rowIndex } = props;

  switch (column.type) {
    case ColumnType.AMOUNT: {
      return (
        <AmountCell<T>
          column={column}
          forcedValue={extra?.computeTotal && extra.computeTotal(data, column.dataIndex)}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.NUMBER: {
      return (
        <NumberCell<T>
          column={column}
          forcedValue={extra?.computeTotal && extra.computeTotal(data, column.dataIndex)}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.PERCENTAGE: {
      return (
        <PercentageCell<T>
          column={column}
          forcedValue={extra?.computeTotal && extra.computeTotal(data, column.dataIndex)}
          extra={extra}
          rowIndex={rowIndex}
        />
      );
    }
    case ColumnType.BADGE:
    case ColumnType.BUTTON:
    case ColumnType.CODE:
    case ColumnType.CUSTOM:
    case ColumnType.DATE:
    case ColumnType.TEXT:
      return <td className='table--footer-value center'>-</td>;

    default: {
      throw new Error('Missing ColumnType');
    }
  }

  throw new Error('Should have returned by then');
};

export default StaticDataTableFooterCell;
