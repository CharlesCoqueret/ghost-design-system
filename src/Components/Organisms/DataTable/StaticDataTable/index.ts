export { default as StaticDataTable } from './StaticDataTable';
