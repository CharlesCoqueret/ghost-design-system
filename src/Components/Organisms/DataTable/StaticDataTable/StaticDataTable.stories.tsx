import React, { ReactElement } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import StaticDataTable, { IStaticDataTableProps } from './StaticDataTable';
import { ColumnType, IColumnType } from '../Common/types';
import { Icon } from '../../../Atoms';

export default {
  title: 'Organism/DataTable/StaticDataTable',
  component: StaticDataTable,
  parameters: { actions: { argTypesRegex: '^on.*' }, controls: { sort: 'requiredFirst' } },
} as ComponentMeta<typeof StaticDataTable>;

const columns: IColumnType<IDemoType>[] = [
  {
    title: 'Code',
    dataIndex: 'id',
    sorter: true,
    type: ColumnType.CODE,
    width: '70px',
  },
  {
    title: 'Text',
    dataIndex: 'name',
    type: ColumnType.TEXT,
    ellipsis: true,
    sorter: true,
    width: '20%',
  },
  {
    title: 'Badge',
    dataIndex: 'status',
    sorter: true,
    type: ColumnType.BADGE,
    options: [
      { label: 'Inactive', value: 'INACTIVE' },
      { label: 'In progress', value: 'IN_PROGRESS' },
      { label: 'Active', value: 'ACTIVE' },
    ],
  },
  {
    title: 'Amount',
    dataIndex: 'price',
    sorter: true,
    type: ColumnType.AMOUNT,
    currency: 'â‚¬',
  },
  {
    title: 'Percentage',
    dataIndex: 'parts',
    sorter: true,
    type: ColumnType.PERCENTAGE,
  },
  {
    title: 'Date',
    dataIndex: 'startDate',
    sorter: true,
    type: ColumnType.DATE,
  },
  {
    title: 'Actions',
    type: ColumnType.BUTTON,
    moreActionsMessage: 'Nore actions',
    buttons: [
      {
        icon: ['fal', 'arrow-to-bottom'],
        label: 'Download',
        onClick: (row) => {
          console.log(`download icon clicked on row: ${row.id}`);
        },
      },
      {
        icon: ['fal', 'eye'],
        label: 'Watch',
        onClick: (row) => {
          console.log(`Eye icon clicked on row: ${row.id}`);
        },
      },
    ],
  },
];

interface IDemoType {
  id: string;
  name: string;
  status: string;
  price: number;
  parts: number;
  startDate: Date;
}

const data = [
  {
    id: 'UGA',
    name: 'Lorem ipsum dolor sit amet consectetur adipiscing elit',
    status: 'INACTIVE',
    price: -100000,
    parts: -10,
    startDate: new Date(2021, 2, 24),
  },
  {
    id: 'ARE',
    name: 'Consecte',
    status: 'ACTIVE',
    price: -2500000,
    parts: 15,
    startDate: new Date(2021, 1, 13),
  },
  {
    id: 'GBR',
    name: 'Sed',
    status: 'IN_PROGRESS',
    price: 150000,
    parts: 3,
    startDate: new Date(2021, 1, 4),
  },
  {
    id: 'JUF',
    name: 'dolore eu fugiat nulla pariatur',
    status: 'IN_PROGRESS',
    price: -354000,
    parts: 70,
    startDate: new Date(2021, 1, 15),
  },
  {
    id: 'YAI',
    name: 'Sed ut perspiciatis',
    status: 'ACTIVE',
    price: 4150000,
    parts: 30,
    startDate: new Date(2021, 2, 17),
  },
  {
    id: 'SML',
    name: 'Ut enim ad minima veniam',
    status: 'ACTIVE',
    price: -70500,
    parts: 10,
    startDate: new Date(2021, 3, 10),
  },
];

const Template: ComponentStory<(args: IStaticDataTableProps<IDemoType>) => ReactElement> = (
  args: IStaticDataTableProps<IDemoType>,
) => {
  return <StaticDataTable<IDemoType> {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  data: data,
  columns: columns,
};

export const ClickableRow = Template.bind({});
ClickableRow.args = {
  data: data,
  columns: columns,
  extra: {
    onRowClick: (row: IDemoType) => {
      console.log(`clicked row: ${row.id}`);
    },
  },
};

export const SelectableRows = Template.bind({});
SelectableRows.args = {
  data: data,
  columns: columns,
  extra: {
    onRowSelect: (rows: Array<IDemoType>, row: IDemoType) => {
      console.log(`Number of rows selected: ${rows.length}\nClicked row: ${row.id}`);
    },
    isSelectable: (row: IDemoType) => {
      return row.id !== 'UGA';
    },
  },
};

export const ComputeTotal = Template.bind({});
ComputeTotal.args = {
  data: data,
  columns: columns,
  extra: {
    computeTotal: (data: Array<IDemoType>, dataIndex: keyof IDemoType) => {
      if (dataIndex === 'price' || dataIndex === 'parts') {
        return data
          .map((row: IDemoType) => {
            return row[dataIndex];
          })
          .reduce((a, b) => Number(a) + Number(b), 0);
      } else {
        return '-';
      }
    },
  },
};

export const SelectableAndComputeTotal = Template.bind({});
SelectableAndComputeTotal.args = {
  data: data,
  columns: columns,
  extra: {
    onRowSelect: (rows: Array<IDemoType>, row: IDemoType) => {
      console.log(`Number of rows selected: ${rows.length}\nClicked row: ${row.id}`);
    },
    computeTotal: (data: Array<IDemoType>, dataIndex: keyof IDemoType) => {
      if (dataIndex === 'price' || dataIndex === 'parts') {
        return data
          .map((row: IDemoType) => {
            return row[dataIndex];
          })
          .reduce((a, b) => Number(a) + Number(b), 0);
      } else {
        return '-';
      }
    },
  },
};

export const NoData = Template.bind({});
NoData.args = {
  data: [],
  columns: columns,
  extra: {
    onRowSelect: (rows: Array<IDemoType>, row: IDemoType) => {
      console.log(`Number of rows selected: ${rows.length}\nClicked row: ${row.id}`);
    },
  },
};

export const Loading = Template.bind({});
Loading.args = {
  data: data,
  columns: columns,
  loading: (
    <div style={{ display: 'flex', justifyContent: 'space-around' }}>
      <Icon icon={['fal', 'spinner']} size='3x' spin />
    </div>
  ),
  extra: {
    onRowSelect: (rows: Array<IDemoType>, row: IDemoType) => {
      console.log(`Number of rows selected: ${rows.length}\nClicked row: ${row.id}`);
    },
    computeTotal: (data: Array<IDemoType>, dataIndex: keyof IDemoType) => {
      if (dataIndex === 'price' || dataIndex === 'parts') {
        return data
          .map((row: IDemoType) => {
            return row[dataIndex];
          })
          .reduce((a, b) => Number(a) + Number(b), 0);
      } else {
        return '-';
      }
    },
  },
};
