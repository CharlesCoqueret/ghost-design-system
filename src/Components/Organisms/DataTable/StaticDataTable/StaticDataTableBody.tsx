import React, { ReactElement, MouseEvent, useState } from 'react';
import classnames from 'classnames';

import StaticDataTableCell from './StaticDataTableCell';
import { IColumnType, IExtraLineEditableInPlaceDataTableProps } from '../Common/types';
import DataTableCellSelectable from '../Common/DataTableCellSelectable';
import { Icon } from '../../../Atoms/Icon';

interface IStaticDataTableBodyProps<T> {
  columns: Array<IColumnType<T>>;
  data: Array<T>;
  extra?: IExtraLineEditableInPlaceDataTableProps<T>;
  loading?: ReactElement;
}

const StaticDataTableBody = <T,>(props: IStaticDataTableBodyProps<T>): ReactElement => {
  const { columns, data, extra, loading } = props;

  const [selectedRows, setSelectedRows] = useState<Record<number, boolean>>({});

  const isSelectable = extra?.onRowSelect;
  const isExtended = extra?.onRowSelect || extra?.computeTotal;

  const handleRowClick = (row: T, rowIndex: number) => {
    return extra && extra.onRowClick
      ? (event: MouseEvent<HTMLElement>) => {
          event.preventDefault();
          if (extra && extra.onRowClick) {
            extra.onRowClick(row, rowIndex);
          }
        }
      : undefined;
  };

  /** Handle */
  const handleSelectClick = (row: T, rowIndex: number) => {
    return (_event: MouseEvent<HTMLElement>, selected: boolean) => {
      const newSelectedRows = { ...selectedRows };
      newSelectedRows[rowIndex] = selected;
      setSelectedRows(newSelectedRows);
      if (extra?.onRowSelect) {
        extra.onRowSelect(
          data.filter((_row, index) => newSelectedRows[index]),
          row,
          rowIndex,
        );
      }
    };
  };

  return (
    <tbody>
      {data.map((row, rowIndex) => {
        return (
          <tr
            key={`row-${rowIndex}`}
            onClick={handleRowClick(row, rowIndex)}
            className={classnames({ pointer: extra && extra.onRowClick, selected: selectedRows[rowIndex] })}>
            {isExtended &&
              (isSelectable ? (
                <DataTableCellSelectable
                  handleSelectClick={handleSelectClick(row, rowIndex)}
                  selected={selectedRows[rowIndex]}
                  selectable={
                    (extra && extra.isSelectable ? extra?.isSelectable(row, rowIndex) : true) && !extra?.editedRowIndex
                  }
                />
              ) : (
                <td></td>
              ))}

            {columns.map((column) => {
              return (
                <StaticDataTableCell<T>
                  key={`cell-${rowIndex}-${column.title}`}
                  column={column}
                  row={row}
                  extra={extra}
                  rowIndex={rowIndex}
                />
              );
            })}
          </tr>
        );
      })}
      {(!data || data?.length === 0) && (
        <tr className='no-data'>
          <td colSpan={columns.length + (isSelectable ? 1 : 0)}>
            <div className='no-data-container'>
              <div className='no-data-text'>{extra?.localization?.noData ?? 'No data'}</div>
              <Icon icon={['fal', 'inbox']} size='2x' className='no-data-icon' />
            </div>
          </td>
        </tr>
      )}
      {loading && (
        <tr className='no-data'>
          <td colSpan={columns.length + (isSelectable ? 1 : 0)}>{loading}</td>
        </tr>
      )}
    </tbody>
  );
};

export default StaticDataTableBody;
