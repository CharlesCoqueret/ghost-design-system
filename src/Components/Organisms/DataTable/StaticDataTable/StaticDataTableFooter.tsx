import React, { ReactElement } from 'react';

import StaticDataTableFooterCell from './StaticDataTableFooterCell';
import { IColumnType, IExtraStaticDataTableProps } from '../Common/types';

interface IStaticDataTableFooterProps<T> {
  columns: Array<IColumnType<T>>;
  data: Array<T>;
  extra?: IExtraStaticDataTableProps<T>;
}

const StaticDataTableFooter = <T,>(props: IStaticDataTableFooterProps<T>): ReactElement => {
  const { columns, data, extra } = props;

  if (!extra?.computeTotal) {
    return <></>;
  }

  return (
    <tfoot>
      <tr>
        <td key='footer-selectable' className='table--footer--selectable'>
          {extra.localization?.total ?? 'Total'}
        </td>

        {columns.map((column) => {
          return (
            <StaticDataTableFooterCell
              column={column}
              data={data}
              extra={extra}
              key={`footer-${column.title}`}
              rowIndex={data.length + 1}
            />
          );
        })}
      </tr>
    </tfoot>
  );
};

export default StaticDataTableFooter;
