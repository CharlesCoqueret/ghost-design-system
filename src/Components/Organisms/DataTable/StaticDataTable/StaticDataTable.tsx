import React, { ReactElement, useCallback, useState } from 'react';

import StaticDataTableBody from './StaticDataTableBody';
import StaticDataTableFooter from './StaticDataTableFooter';
import StaticDataTableHeader from './StaticDataTableHeader';
import { IColumnType, IExtraStaticDataTableProps, SortDirectionEnum } from '../Common/types';

export interface IStaticDataTableProps<T> {
  data: Array<T>;
  columns: Array<IColumnType<T>>;
  extra?: IExtraStaticDataTableProps<T>;
  loading?: ReactElement;
  onSortChange?: (sortField?: keyof T, sortDirection?: SortDirectionEnum) => void;
}

const StaticDataTable = <T,>(props: IStaticDataTableProps<T>): ReactElement => {
  const { data, columns, extra, loading, onSortChange } = props;

  const [sortField, setSortField] = useState<keyof T | undefined>();
  const [sortDirection, setSortDirection] = useState<SortDirectionEnum | undefined>();

  const handleSortChange = useCallback((newSortField: keyof T, newSortDirection?: SortDirectionEnum) => {
    if (sortField !== newSortField || newSortDirection !== newSortDirection) {
      setSortField(newSortField);
      setSortDirection(newSortDirection);

      if (onSortChange) {
        if (newSortField && newSortDirection) onSortChange(newSortField, newSortDirection);
        else onSortChange();
      }
    }
  }, []);

  return (
    <table className='gds-table'>
      <StaticDataTableHeader<T>
        columns={columns}
        extra={extra}
        onSortChange={handleSortChange}
        sortDirection={sortDirection}
        sortField={sortField}
      />
      <StaticDataTableBody<T> columns={columns} data={data} extra={extra} loading={loading} />
      <StaticDataTableFooter<T> columns={columns} data={data} extra={extra} />
    </table>
  );
};

StaticDataTable.defaultProps = {
  columns: [],
  data: [],
};

export default StaticDataTable;
