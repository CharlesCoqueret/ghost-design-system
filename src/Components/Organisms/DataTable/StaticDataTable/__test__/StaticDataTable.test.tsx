import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Mocking suneditor which is problematic with Jest
jest.mock('suneditor', () => {});
jest.mock('suneditor/src/plugins/', () => {});
jest.mock('suneditor/src/plugins/submenu/align', () => {});
jest.mock('suneditor/src/plugins/command/blockquote', () => {});
jest.mock('suneditor/src/plugins/submenu/fontColor', () => {});
jest.mock('suneditor/src/plugins/submenu/fontSize', () => {});
jest.mock('suneditor/src/plugins/submenu/formatBlock', () => {});
jest.mock('suneditor/src/plugins/submenu/hiliteColor', () => {});
jest.mock('suneditor/src/plugins/submenu/horizontalRule', () => {});
jest.mock('suneditor/src/plugins/dialog/image', () => {});
jest.mock('suneditor/src/plugins/dialog/link', () => {});
jest.mock('suneditor/src/plugins/submenu/lineHeight', () => {});
jest.mock('suneditor/src/plugins/submenu/list', () => {});
jest.mock('suneditor/src/plugins/submenu/paragraphStyle', () => {});
jest.mock('suneditor/src/plugins/submenu/table', () => {});
jest.mock('suneditor-react', () => {});
jest.mock('suneditor-react/dist', () => {});
jest.mock('suneditor-react/dist/types/lang', () => {});

import StaticDataTable from '../StaticDataTable';
import { ColumnType, SortDirectionEnum } from '../../Common/types';

describe('StaticDataTable component', () => {
  it('StaticDataTable renders and handles sort', () => {
    const onSortChangeMock = jest.fn();

    const { container } = render(
      <StaticDataTable<{ number: number }>
        columns={[
          {
            dataIndex: 'number',
            dataTestId: 'DATA-TEST-ID-number',
            sorter: true,
            title: 'Number',
            type: ColumnType.NUMBER,
          },
        ]}
        data={[{ number: 1 }]}
        onSortChange={onSortChangeMock}
      />,
    );

    expect(container).toMatchSnapshot();

    const numberSort = screen.getByTestId('DATA-TEST-ID-number-sort');
    userEvent.click(numberSort);

    expect(container).toMatchSnapshot();
    expect(onSortChangeMock).toBeCalledTimes(1);
    expect(onSortChangeMock).toBeCalledWith('number', SortDirectionEnum.DESC);

    userEvent.click(numberSort);

    expect(container).toMatchSnapshot();
    expect(onSortChangeMock).toBeCalledTimes(2);
    expect(onSortChangeMock).toBeCalledWith('number', SortDirectionEnum.ASC);

    userEvent.click(numberSort);

    expect(container).toMatchSnapshot();
    expect(onSortChangeMock).toBeCalledTimes(3);
    expect(onSortChangeMock).toBeCalledWith();
  });
});
