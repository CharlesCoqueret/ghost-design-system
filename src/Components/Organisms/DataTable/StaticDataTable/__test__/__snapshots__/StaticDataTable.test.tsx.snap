// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticDataTable component StaticDataTable renders and handles sort 1`] = `
<div>
  <table
    class="gds-table"
  >
    <thead>
      <tr>
        <th
          class=""
        >
          <div
            class="table--header-value table--header-value--sortable"
            data-testid="DATA-TEST-ID-number-sort"
            tabindex="0"
          >
            <span
              class="gds-typography table--header-value--title right ellipsis"
              title="Number"
            >
              Number
            </span>
            <span
              class="table--header-value--sorter"
            >
              <i
                class="icon"
                icon="fas,sort"
              />
            </span>
          </div>
          <div>
            Click to sort
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
        tabindex="-1"
      >
        <td
          class=""
        >
          <div
            class="gds-field-group"
          >
            <span
              class="gds-amount-field amount-field-read-only table--cell-value--number-readonly"
              id="number"
              name="number"
            >
              1
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTable component StaticDataTable renders and handles sort 2`] = `
<div>
  <table
    class="gds-table"
  >
    <thead>
      <tr>
        <th
          class=""
        >
          <div
            class="table--header-value table--header-value--sortable"
            data-testid="DATA-TEST-ID-number-sort"
            tabindex="0"
          >
            <span
              class="gds-typography table--header-value--title right ellipsis"
              title="Number"
            >
              Number
            </span>
            <span
              class="table--header-value--sorter"
            >
              <i
                class="icon"
                icon="fas,sort-down"
              />
            </span>
          </div>
          <div>
            Click to sort
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
        tabindex="-1"
      >
        <td
          class=""
        >
          <div
            class="gds-field-group"
          >
            <span
              class="gds-amount-field amount-field-read-only table--cell-value--number-readonly"
              id="number"
              name="number"
            >
              1
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTable component StaticDataTable renders and handles sort 3`] = `
<div>
  <table
    class="gds-table"
  >
    <thead>
      <tr>
        <th
          class=""
        >
          <div
            class="table--header-value table--header-value--sortable"
            data-testid="DATA-TEST-ID-number-sort"
            tabindex="0"
          >
            <span
              class="gds-typography table--header-value--title right ellipsis"
              title="Number"
            >
              Number
            </span>
            <span
              class="table--header-value--sorter"
            >
              <i
                class="icon"
                icon="fas,sort-up"
              />
            </span>
          </div>
          <div>
            Click to sort
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
        tabindex="-1"
      >
        <td
          class=""
        >
          <div
            class="gds-field-group"
          >
            <span
              class="gds-amount-field amount-field-read-only table--cell-value--number-readonly"
              id="number"
              name="number"
            >
              1
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`StaticDataTable component StaticDataTable renders and handles sort 4`] = `
<div>
  <table
    class="gds-table"
  >
    <thead>
      <tr>
        <th
          class=""
        >
          <div
            class="table--header-value table--header-value--sortable"
            data-testid="DATA-TEST-ID-number-sort"
            tabindex="0"
          >
            <span
              class="gds-typography table--header-value--title right ellipsis"
              title="Number"
            >
              Number
            </span>
            <span
              class="table--header-value--sorter"
            >
              <i
                class="icon"
                icon="fas,sort"
              />
            </span>
          </div>
          <div>
            Click to sort
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
        tabindex="-1"
      >
        <td
          class=""
        >
          <div
            class="gds-field-group"
          >
            <span
              class="gds-amount-field amount-field-read-only table--cell-value--number-readonly"
              id="number"
              name="number"
            >
              1
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
